src/Grpc.Net.Client/Balancer/SubchannelsLoadBalancer.cs

```diff
@@ -0,0 +1,296 @@
+        /// <param name="loggerFactory">The logger factory.</param>
+        protected SubchannelsLoadBalancer(IChannelControlHelper controller, ILoggerFactory loggerFactory)
+        {
+            _addressSubchannels = new List<AddressSubchannel>();
+            Controller = controller;
+            _logger = loggerFactory.CreateLogger(GetType());
+        }
+
+        private void ResolverError(Status status)
+        {
+            // If balancer doesn't have a ready subchannel then remove any current subchannels
+            // and update channel state with resolver error.
+            switch (State)
+            {
+                case ConnectivityState.Idle:
+                case ConnectivityState.Connecting:
+                case ConnectivityState.TransientFailure:
+                    foreach (var addressSubchannel in _addressSubchannels)
+                    {
+                        RemoveSubchannel(addressSubchannel.Subchannel);
+                    }
+                    Controller.UpdateState(new BalancerState(ConnectivityState.TransientFailure, new ErrorPicker(status)));
+                    break;
+            }
+        }
+
+        private int? FindSubchannelByAddress(List<AddressSubchannel> addressSubchannels, DnsEndPoint endPoint)
+        {
+            for (var i = 0; i < addressSubchannels.Count; i++)
+            {
+                var s = addressSubchannels[i];
+                if (Equals(s.Address, endPoint))
+                {
+                    return i;
+                }
+            }
+
+            return null;
+        }
+
+        private int? FindSubchannel(List<AddressSubchannel> addressSubchannels, Subchannel subchannel)
+        {
+            for (var i = 0; i < addressSubchannels.Count; i++)
+            {
+                var s = addressSubchannels[i];
+                if (Equals(s.Subchannel, subchannel))
+                {
+                    return i;
+                }
+            }
+
+            return null;
+        }
+
+        /// <inheritdoc />
+        public override void UpdateChannelState(ChannelState state)
+        {
+            if (state.Status.StatusCode != StatusCode.OK)
+            {
+                ResolverError(state.Status);
+                return;
+            }
+            if (state.Addresses == null || state.Addresses.Count == 0)
+            {
+                ResolverError(new Status(StatusCode.Unavailable, "Resolver returned no addresses."));
+                return;
+            }
+
+            var allUpdatedSubchannels = new List<AddressSubchannel>();
+            var newSubchannels = new List<Subchannel>();
+            var currentSubchannels = _addressSubchannels.ToList();
+
+            foreach (var address in state.Addresses)
+            {
+                var i = FindSubchannelByAddress(currentSubchannels, address);
+
+                AddressSubchannel newSubConnection;
+                if (i != null)
```

Comment:
I think I'm a little confused here. When updating state, all current subchannels that match with the addresses on the state are removed and new subchannels are created for new addresses. Why are the subchannels for current addresses not recreated after being removed?

Reply:
Added more comments.